File: connectors/java/src/main/java/com/mochilink/connector/subscription/SubscriptionManager.java
package com.mochilink.connector.subscription;

import java.util.Collection;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;
import java.util.logging.Logger;

/**
 * Manages event subscriptions for the connector
 * 
 * Handles subscription lifecycle, filtering, and matching logic
 * to determine which events should be sent to the management server.
 */
public class SubscriptionManager {
    
    private final Logger logger;
    private final Map<String, EventSubscription> subscriptions;
    
    public SubscriptionManager(Logger logger) {
        this.logger = logger;
        this.subscriptions = new ConcurrentHashMap<>();
    }
    
    /**
     * Add a new subscription
     */
    public void addSubscription(String subscriptionId, EventSubscription subscription) {
        subscriptions.put(subscriptionId, subscription);
        logger.info(String.format("Added subscription: %s for events: %s", 
                                 subscriptionId, subscription.getEventTypes()));
    }
    
    /**
     * Remove a subscription
     */
    public void removeSubscription(String subscriptionId) {
        EventSubscription removed = subscriptions.remove(subscriptionId);
        if (removed != null) {
            logger.info(String.format("Removed subscription: %s", subscriptionId));
        } else {
            logger.warning(String.format("Subscription not found: %s", subscriptionId));
        }
    }
    
    /**
     * Check if there are any subscriptions for the given event type
     */
    public boolean hasSubscription(String eventType) {
        return subscriptions.values().stream()
            .anyMatch(sub -> sub.includesEventType(eventType));
    }
    
    /**
     * Check if event data matches any subscription filters for the given event type
     */
    public boolean matchesFilters(String eventType, Map<String, Object> eventData) {
        return subscriptions.values().stream()
            .filter(sub -> sub.includesEventType(eventType))
            .anyMatch(sub -> sub.matchesFilters(eventData));
    }
    
    /**
     * Get a specific subscription by ID
     */
    public EventSubscription getSubscription(String subscriptionId) {
        return subscriptions.get(subscriptionId);
    }
    
    /**
     * Get all subscriptions
     */
    public Collection<EventSubscription> getAllSubscriptions() {
        return subscriptions.values();
    }
    
    /**
     * Get subscription count
     */
    public int getSubscriptionCount() {
        return subscriptions.size();
    }
    
    /**
     * Clear all subscriptions
     */
    public void clearAll() {
        int count = subscriptions.size();
        subscriptions.clear();
        logger.info(String.format("Cleared %d subscriptions", count));
    }
    
    /**
     * Check if a subscription exists
     */
    public boolean hasSubscriptionId(String subscriptionId) {
        return subscriptions.containsKey(subscriptionId);
    }
}

File: mochi-link-connectors/bedrock-edition/pmmp/src/com/mochilink/connector/pmmp/subscription/SubscriptionManager.php
<?php

declare(strict_types=1);

namespace com\mochilink\connector\pmmp\subscription;

use pocketmine\plugin\PluginLogger;

/**
 * Manages event subscriptions for the PMMP connector
 * 
 * Handles subscription lifecycle, filtering, and matching logic
 * to determine which events should be sent to the management server.
 * 
 * @author chm413
 * @version 1.0.0
 */
class SubscriptionManager {
    
    private PluginLogger $logger;
    /** @var EventSubscription[] */
    private array $subscriptions = [];
    
    public function __construct(PluginLogger $logger) {
        $this->logger = $logger;
    }
    
    /**
     * Add a new subscription
     */
    public function addSubscription(string $subscriptionId, EventSubscription $subscription): void {
        $this->subscriptions[$subscriptionId] = $subscription;
        $this->logger->info(sprintf(
            "Added subscription: %s for events: %s",
            $subscriptionId,
            json_encode($subscription->getEventTypes())
        ));
    }
    
    /**
     * Remove a subscription
     */
    public function removeSubscription(string $subscriptionId): void {
        if (isset($this->subscriptions[$subscriptionId])) {
            unset($this->subscriptions[$subscriptionId]);
            $this->logger->info(sprintf("Removed subscription: %s", $subscriptionId));
        } else {
            $this->logger->warning(sprintf("Subscription not found: %s", $subscriptionId));
        }
    }
    
    /**
     * Check if there are any subscriptions for the given event type
     */
    public function hasSubscription(string $eventType): bool {
        foreach ($this->subscriptions as $subscription) {
            if ($subscription->includesEventType($eventType)) {
                return true;
            }
        }
        return false;
    }
    
    /**
     * Check if event data matches any subscription filters for the given event type
     */
    public function matchesFilters(string $eventType, array $eventData): bool {
        foreach ($this->subscriptions as $subscription) {
            if ($subscription->includesEventType($eventType) && $subscription->matchesFilters($eventData)) {
                return true;
            }
        }
        return false;
    }
    
    /**
     * Get a specific subscription by ID
     */
    public function getSubscription(string $subscriptionId): ?EventSubscription {
        return $this->subscriptions[$subscriptionId] ?? null;
    }
    
    /**
     * Get all subscriptions
     */
    public function getAllSubscriptions(): array {
        return array_values($this->subscriptions);
    }
    
    /**
     * Get subscription count
     */
    public function getSubscriptionCount(): int {
        return count($this->subscriptions);
    }
    
    /**
     * Clear all subscriptions
     */
    public function clearAll(): void {
        $count = count($this->subscriptions);
        $this->subscriptions = [];
        $this->logger->info(sprintf("Cleared %d subscriptions", $count));
    }
    
    /**
     * Check if a subscription exists
     */
    public function hasSubscriptionId(string $subscriptionId): bool {
        return isset($this->subscriptions[$subscriptionId]);
    }
}
